### PR TITLE
Implementation of the fragment of pokemon types

### PR DESCRIPTION
**Pull request by ezequielsetzesglobant**
The feature of displaying a list of the types of pokemons in one of the tabs was implemented.

Pixel
https://user-images.githubusercontent.com/84338513/136972277-9470a859-43de-4531-b216-89e220f2c675.mp4

Nexus
https://user-images.githubusercontent.com/84338513/137028613-c40c183b-f186-4714-987f-483d483c0a10.mp4

Unit test of ViewModels, Mappers and UseCase
image
image

Coverage of ViewModels, Mappers and UseCase
image
image
image

### CODE REVIEW

File: app/src/main/java/com/example/pokemonapp/fragment/CardTypeFragment.kt [hunk 1]
```diff
@@ -5,14 +5,54 @@ import android.view.LayoutInflater
...
 class CardTypeFragment : Fragment() {
 
+    private lateinit var adapter: TypeAdapter
     private lateinit var binding: FragmentCardTypeBinding
+    private val viewModel by viewModel<CardTypeViewModel>()
 
-    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
+    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
         binding = FragmentCardTypeBinding.inflate(inflater, container, false)
+        viewModel.getLiveDataState().observe(viewLifecycleOwner, updateUIObserver)
```
Comment (BQuinteros) [inline app/src/main/java/com/example/pokemonapp/fragment/CardTypeFragment.kt:29]: I think this methods should be in the on view created not when the view is create

File: app/src/main/java/com/example/pokemonapp/fragment/CardTypeFragment.kt [hunk 2]
```diff
@@ -5,14 +5,54 @@ import android.view.LayoutInflater
...
+                binding.textViewCardTypeFragmentTypesError.text = eventContent.error
+            }
+            CardTypeViewModel.CardTypeStatus.LOADING -> {
+                binding.progressBarCardTypeFragmentLoad.visibility = View.VISIBLE
+            }
+        }
+    }
+
+    private fun initializeRecyclerView() {
+        binding.recyclerViewCardTypeFragmentTypesList.layoutManager = GridLayoutManager(context, COLUMS_AMOUNT)
```
Comment (BQuinteros) [inline app/src/main/java/com/example/pokemonapp/fragment/CardTypeFragment.kt]: here you can do something like this 
```
binding.reclyclerviewCardTypeFragmment.apply {
    layoutManager = GridLayoutManager(context, COLUMS_AMOUNT)    
    adapter = TypeAdapter(emptyList())
}
```
but why you first gettypes and then you initialize the adapter

File: app/build.gradle [hunk 3]
```diff
@@ -40,8 +40,6 @@ android {
 preBuild.dependsOn ktlint
 
 dependencies {
-    implementation "org.jetbrains.kotlin:kotlin-stdlib:$kotlin_version"
```
Comment (SartiMau) [inline app/build.gradle:43]: is this okay? why did you delete kotlin from the app module?

File: app/src/main/java/com/example/pokemonapp/fragment/CardTypeFragment.kt [hunk 4]
```diff
@@ -5,14 +5,54 @@ import android.view.LayoutInflater
...
         binding = FragmentCardTypeBinding.inflate(inflater, container, false)
+        viewModel.getLiveDataState().observe(viewLifecycleOwner, updateUIObserver)
+        viewModel.getTypes()
+        initializeRecyclerView()
         return binding.root
     }
+
+    private val updateUIObserver = Observer<Event<CardTypeViewModel.CardTypeData>> { event ->
+        val eventContent = event.getContentIfNotHandled()
+        when (eventContent?.responseType) {
```
Comment (SartiMau) [inline app/src/main/java/com/example/pokemonapp/fragment/CardTypeFragment.kt:35]: please unify this line with the above one

File: app/src/main/java/com/example/pokemonapp/adapter/TypeAdapter.kt [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+    class TypeHolder(private val binding: ItemRecyclerViewCardTypeFragmentBinding, private val context: Context) :
+        RecyclerView.ViewHolder(binding.root) {
+
+        fun bind(type: String) {
+            binding.textViewItemRecyclerViewCardTypeFragmentType.text =
+                context.getString(R.string.text_view_item_recycler_view_card_type_fragment_type, type)
+            binding.imageViewItemRecyclerViewCardTypeFragmentType.setImageResource(ImageType.imageMapper(type))
+        }
+
+        enum class ImageType(val type: String) {
```
Comment (SartiMau) [inline app/src/main/java/com/example/pokemonapp/adapter/TypeAdapter.kt]: I think that will be better if you put this enum in a separated file, maybe in a utils package

File: app/src/main/java/com/example/pokemonapp/fragment/CardTypeFragment.kt [hunk 6]
```diff
@@ -5,14 +5,54 @@ import android.view.LayoutInflater
...
+                binding.progressBarCardTypeFragmentLoad.visibility = View.GONE
+                binding.textViewCardTypeFragmentTypesError.text = eventContent.error
+            }
+            CardTypeViewModel.CardTypeStatus.LOADING -> {
+                binding.progressBarCardTypeFragmentLoad.visibility = View.VISIBLE
+            }
+        }
+    }
+
+    private fun initializeRecyclerView() {
```
Comment (SartiMau) [inline app/src/main/java/com/example/pokemonapp/fragment/CardTypeFragment.kt]: please, do not initialize the recycler empty, you can initialize it when you get all the info in the SUCCESSFUL state

File: app/src/main/res/layout/fragment_card_type.xml [hunk 7]
```diff
@@ -1,14 +1,33 @@
...
         android:layout_width="match_parent"
         android:layout_height="match_parent"
-        android:gravity="center"
-        android:text="@string/text_view_card_type_fragment_text" />
+        android:gravity="center" />
+
+    <androidx.recyclerview.widget.RecyclerView
+        android:id="@+id/recycler_view_card_type_fragment_types_list"
+        android:layout_width="match_parent"
+        android:layout_height="@dimen/recycler_view_card_type_fragment_height"
```
Comment (SartiMau) [inline app/src/main/res/layout/fragment_card_type.xml]: it is really necessary with height with a dimen? can't you use it like 0dp?

File: app/src/main/res/values/strings.xml [hunk 8]
```diff
@@ -4,7 +4,13 @@
     <string name="tab_layout_main_fragment_tab_name_type">Pokemon Types</string>
     <string name="tab_layout_main_fragment_tab_name_subtypes">Pokemon Subtypes</string>
     <string name="tab_layout_main_fragment_tab_name_supertypes">Pokemon Supertypes</string>
-    <string name="text_view_card_type_fragment_text">Types Fragment</string>
-    <string name="text_view_card_sub_type_fragment_text">Subtypes Fragment</string>
-    <string name="text_view_card_super_type_fragment_text">Supertypes Fragment</string>
+
+    <string name="text_view_card_type_fragment_text">%1$s</string>
```
Comment (SartiMau) [inline app/src/main/res/values/strings.xml]: you have this value in 4 different strings, maybe you can create a general, text_view_one_parameter and reuse it

**Review by SartiMau — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/example/pokemonapp/fragment/CardTypeFragment.kt [hunk 9]
```diff
@@ -5,14 +5,54 @@ import android.view.LayoutInflater
...
+    private lateinit var adapter: TypeAdapter
     private lateinit var binding: FragmentCardTypeBinding
+    private val viewModel by viewModel<CardTypeViewModel>()
 
-    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
+    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
         binding = FragmentCardTypeBinding.inflate(inflater, container, false)
+        viewModel.getLiveDataState().observe(viewLifecycleOwner, updateUIObserver)
+        viewModel.getTypes()
+        initializeRecyclerView()
```
Comment (nicolbidone) [inline app/src/main/java/com/example/pokemonapp/fragment/CardTypeFragment.kt]: you should set this in OnViewCreated

File: data/src/main/java/com/example/data/database/PokemonDataBase.kt [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+
+@Database(entities = [TypeEntity::class], version = 1)
+@TypeConverters(ListConverter::class)
+abstract class PokemonDataBase : RoomDatabase() {
+
+    abstract fun pokemonDao(): PokemonDao
+
+    companion object {
+
+        private var INSTANCE: PokemonDataBase? = null
```
Comment (nicolbidone) [inline data/src/main/java/com/example/data/database/PokemonDataBase.kt]: var name should be in low case

File: app/src/main/java/com/example/pokemonapp/fragment/CardTypeFragment.kt (hunk 1 above)
Comment (ezequielsetzesglobant) [reply app/src/main/java/com/example/pokemonapp/fragment/CardTypeFragment.kt:29]: done

File: app/src/main/java/com/example/pokemonapp/fragment/CardTypeFragment.kt (hunk 2 above)
Comment (ezequielsetzesglobant) [reply app/src/main/java/com/example/pokemonapp/fragment/CardTypeFragment.kt]: Those two lines go backwards, there I change them

File: app/build.gradle (hunk 3 above)
Comment (ezequielsetzesglobant) [reply app/build.gradle:43]: Yes, it is ok because it is down the right way from the Dependencies file

File: domain/src/main/java/com/example/domain/entity/TypeData.kt [hunk 11]
```diff
@@ -0,0 +1,5 @@
+package com.example.domain.entity
+
+data class TypeData(
+    val data: List<String> = emptyList(),
+) : java.io.Serializable
```
Comment (nicolbidone) [inline domain/src/main/java/com/example/domain/entity/TypeData.kt]: this should be just Serializable

File: app/src/main/java/com/example/pokemonapp/fragment/CardTypeFragment.kt (hunk 4 above)
Comment (ezequielsetzesglobant) [reply app/src/main/java/com/example/pokemonapp/fragment/CardTypeFragment.kt:35]: I don't do that because on line 39 and 45 I use it too

File: app/src/main/java/com/example/pokemonapp/adapter/TypeAdapter.kt (hunk 5 above)
Comment (ezequielsetzesglobant) [reply app/src/main/java/com/example/pokemonapp/adapter/TypeAdapter.kt]: done

File: app/src/main/res/layout/fragment_card_type.xml (hunk 7 above)
Comment (ezequielsetzesglobant) [reply app/src/main/res/layout/fragment_card_type.xml:18]: that dimension is a 0dp the same, it is like this so as not to have a hardcode

Comment (ezequielsetzesglobant) [reply app/src/main/res/layout/fragment_card_type.xml]: that dimension is a 0dp the same, it is like this so as not to have a hardcode

File: app/src/main/res/values/strings.xml (hunk 8 above)
Comment (ezequielsetzesglobant) [reply app/src/main/res/values/strings.xml]: done

File: app/src/main/java/com/example/pokemonapp/fragment/CardTypeFragment.kt (hunk 6 above)
Comment (ezequielsetzesglobant) [reply app/src/main/java/com/example/pokemonapp/fragment/CardTypeFragment.kt]: done

File: app/src/main/java/com/example/pokemonapp/adapter/TypeAdapter.kt [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+    override fun onBindViewHolder(holder: TypeHolder, position: Int) {
+        holder.bind(types[position])
+    }
+
+    override fun getItemCount(): Int = types.size
+
+    fun setTypes(types: List<String>) {
+        this.types = types
+        notifyDataSetChanged()
+    }
```
Comment (nicolbidone) [inline app/src/main/java/com/example/pokemonapp/adapter/TypeAdapter.kt]: why you don't initialize this in the class constructor?

File: app/src/main/java/com/example/pokemonapp/utils/ImageType.kt [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+            DRAGON -> R.drawable.dragon
+            FAIRY -> R.drawable.fairy
+            FIGHTING -> R.drawable.fighting
+            FIRE -> R.drawable.fire
+            GRASS -> R.drawable.grass
+            LIGHTNING -> R.drawable.lightning
+            METAL -> R.drawable.metal
+            PSYCHIC -> R.drawable.psychic
+            WATER -> R.drawable.water
+            DEFAULT_POKEBALL -> R.drawable.default_pokeball
```
Comment (nicolbidone) [inline app/src/main/java/com/example/pokemonapp/utils/ImageType.kt]: why not to have this in the enums?

File: app/src/main/java/com/example/pokemonapp/fragment/CardTypeFragment.kt (hunk 9 above)
Comment (ezequielsetzesglobant) [reply app/src/main/java/com/example/pokemonapp/fragment/CardTypeFragment.kt]: done

File: data/src/main/java/com/example/data/database/PokemonDataBase.kt (hunk 10 above)
Comment (ezequielsetzesglobant) [reply data/src/main/java/com/example/data/database/PokemonDataBase.kt]: fixed

File: domain/src/main/java/com/example/domain/entity/TypeData.kt (hunk 11 above)
Comment (ezequielsetzesglobant) [reply domain/src/main/java/com/example/domain/entity/TypeData.kt]: fixed

File: app/src/main/java/com/example/pokemonapp/adapter/TypeAdapter.kt (hunk 12 above)
Comment (ezequielsetzesglobant) [reply app/src/main/java/com/example/pokemonapp/adapter/TypeAdapter.kt]: in the constructor there is also: class TypeAdapter (private var types: List <String>)

File: app/src/main/java/com/example/pokemonapp/utils/ImageType.kt (hunk 13 above)
Comment (ezequielsetzesglobant) [reply app/src/main/java/com/example/pokemonapp/utils/ImageType.kt]: I am going to a change it to other thing

**Review by nicolbidone — APPROVED** [review]

File: app/src/main/java/com/example/pokemonapp/fragment/CardTypeFragment.kt [hunk 14]
```diff
@@ -5,14 +5,63 @@ import android.view.LayoutInflater
...
+            }
+        }
+    }
+
+    private fun initializeRecyclerView(types: List<String>) {
+        if (::typeAdapter.isInitialized) {
+            typeAdapter.setTypes(types)
+        } else {
+            typeAdapter = TypeAdapter(types)
+        }
```
Comment (mauroLonghi) [inline app/src/main/java/com/example/pokemonapp/fragment/CardTypeFragment.kt]: mm I think tha if you are initializing the recycler, the adapter always not be inizialized...so this condition its not necessary

Comment (ezequielsetzesglobant) [reply app/src/main/java/com/example/pokemonapp/fragment/CardTypeFragment.kt]: fixed

File: app/src/main/res/layout/fragment_card_type.xml (hunk 7 above)
Comment (SartiMau) [reply app/src/main/res/layout/fragment_card_type.xml]: please put the 0dp because here does not mean "no space" means "all that you can"

Comment (ezequielsetzesglobant) [reply app/src/main/res/layout/fragment_card_type.xml]: done

**Review by SartiMau — APPROVED** [review]

**Review by mauroLonghi — APPROVED** [review]

**Review by BQuinteros — APPROVED** [review]
